lib/srv/reexec.go

```diff
@@ -764,39 +607,208 @@ func runForward(handler forwardHandler) (errw io.Writer, code int, err error) {
 	for {
 		buf := make([]byte, 1024)
 		fbuf := make([]*os.File, 1)
-		n, fn, err := uds.ReadWithFDs(conn, buf, fbuf)
+		n, fn, err := uds.ReadWithFDs(parentConn, buf, fbuf)
 		if err != nil {
 			if utils.IsOKNetworkError(err) {
 				return errorWriter, teleport.RemoteCommandSuccess, nil
 			}
 			return errorWriter, teleport.RemoteCommandFailure, trace.Wrap(err)
 		}
-		addr := string(buf[:n])
-		if fn == 0 {
-			log.Errorf("Parent did not send a file descriptor for address %q.", addr)
-			continue
-		}
 
 		go func() {
-			if err := handler(ctx, addr, fbuf[0]); err != nil {
-				log.WithError(err).Errorf("Error handling forwarding request for address %q.", addr)
+			var req networking.Request
+			if err := json.Unmarshal(buf[:n], &req); err != nil {
+				slog.With("error", err).ErrorContext(ctx, "Error parsing networking request.")
+				return
```

Comment:
This leaks the files in `fbuf[:fn]`; it'd be cleaner if this closure was a function that took a payload and a file, or a payload and a pre-parsed control socket (the "outer" protocol with the json and socketpair doesn't depend on the request, after all).